scheduler/__init__.py

```diff
@@ -1,115 +1,283 @@
+                f"Failed to trigger document fetch for {schedule['companyId']} - {schedule['reportType']}"
+            )
     except Exception as e:
-        logging.error(f"Error in trigger_document_fetch: {str(e)}")
-        schedule['summarized_today'] = False
-        return False
-    
+        logging.error(f"Error in [trigger_document_fetch]: {str(e)}")
+        schedule["summarized_today"] = False
+        return summarization
     finally:
         # Always update last run time, regardless of outcome
-        schedule['lastRun'] = start_time
+        schedule["lastRun"] = start_time
         try:
-            cosmos_data_loader.update_last_run(schedule)
+            schedules_db_manager.update_last_run(schedule)
         except Exception as e:
-            raise LastRunUpdateError(schedule.get('id', 'unknown'), e)
+            raise LastRunUpdateError(schedule.get("id", "unknown"), e)
+
+    return summarization
+
+
+def check_subscription_statuses(orgs: List[Dict]) -> List[Dict]:
+    """Check if the subscription is active and it has financial assistant tier"""
+
+    organizations = []
+
+    if len(orgs) == 0:
+        logger.error("No active organizations found")
+        return organizations
+
+    def check_subscription_status(subscription_id: str) -> dict:
+        try:
+            response = requests.get(
+                STRIPE_SUBSCRIPTION_ENDPOINT.replace(
+                    "<subscription_id>", subscription_id
+                ),
+                timeout=TIMEOUT_SECONDS,
+                headers={
+                    "Content-Type": "application/json",
+                    "X-MS-CLIENT-PRINCIPAL-ID": "00000000-0000-0000-0000-000000000000",
+                },
+            )
+            response_json = response.json()
+
+            if "subscriptionData" not in response_json:
+                raise requests.exceptions.RequestException(
+                    "Subscription data not found in response"
+                )
+
+            if "status" not in response_json["subscriptionData"]:
+                raise requests.exceptions.RequestException(
+                    "Subscription status not found in response"
+                )
+
+            if response_json["subscriptionData"]["status"] != "active":
+                raise requests.exceptions.RequestException("Subscription is not active")
+
+            if "subscriptionTiers" not in response_json:
+                raise requests.exceptions.RequestException(
+                    "Subscription tiers not found in response"
+                )
+
+            return {
+                "subscription_id": subscription_id,
+                "tier": response_json["subscriptionTiers"],
+            }
+
+        except requests.exceptions.RequestException as e:
+            logger.error(
+                f"Failed to check subscription status for {subscription_id}: {str(e)}"
+            )
+            return {"subscription_id": subscription_id, "tier": []}
+
+    for org in orgs:
```

Comment:
@PedroLabrador This sequentially checks each organization’s subscription, which can be slow if there are many organizations

Use multithreading or async requests to speed up execution.
```python
import concurrent.futures

with concurrent.futures.ThreadPoolExecutor() as executor:
    results = executor.map(lambda org: check_subscription_status(org["subscriptionId"]), orgs)

organizations = [
    org for org, status in zip(orgs, results)
    if len(status["tier"]) > 0 and 
       ("Financial Assistant" in status["tier"] or "Premium + Financial Assistant" in status["tier"])
]


```

Reply:
Are you sure you want me to Asynchronously call the endpoints in the backend, to generate reports/send emails, all at the same time?